core/trino-main/src/main/java/io/trino/sql/analyzer/JsonPathAnalyzer.java

```diff
@@ -0,0 +1,533 @@
+                try {
+                    resultType = metadata.resolveFunction(session, QualifiedName.of("abs"), fromTypes(sourceType)).getSignature().getReturnType();
+                }
+                catch (TrinoException e) {
+                    throw semanticException(INVALID_PATH, pathNode, e, "cannot perform JSON path abs() method with %s argument: %s", sourceType.getDisplayName(), e.getMessage());
+                }
+                types.put(PathNodeRef.of(node), resultType);
+                return resultType;
+            }
+
+            return null;
+        }
+
+        @Override
+        protected Type visitArithmeticBinary(ArithmeticBinary node, Void context)
+        {
+            Type leftType = process(node.getLeft());
+            Type rightType = process(node.getRight());
+            if (leftType != null && rightType != null) {
+                BoundSignature signature;
+                try {
+                    signature = metadata.resolveOperator(session, OperatorType.valueOf(node.getOperator().name()), ImmutableList.of(leftType, rightType)).getSignature();
+                }
+                catch (OperatorNotFoundException e) {
+                    throw semanticException(INVALID_PATH, pathNode, e, "invalid operand types (%s and %s) in JSON path arithmetic binary expression: %s", leftType.getDisplayName(), rightType.getDisplayName(), e.getMessage());
+                }
+                Type resultType = signature.getReturnType();
+                types.put(PathNodeRef.of(node), resultType);
+                return resultType;
+            }
+
+            return null;
+        }
+
+        @Override
+        protected Type visitArithmeticUnary(ArithmeticUnary node, Void context)
+        {
+            Type sourceType = process(node.getBase());
+            if (sourceType != null) {
+                if (node.getSign() == PLUS) {
+                    if (!isNumericType(sourceType)) {
+                        throw semanticException(INVALID_PATH, pathNode, "Invalid operand type (%s) in JSON path arithmetic unary expression", sourceType.getDisplayName());
+                    }
+                    types.put(PathNodeRef.of(node), sourceType);
+                    return sourceType;
+                }
+                else {
+                    Type resultType;
+                    try {
+                        resultType = metadata.resolveOperator(session, NEGATION, ImmutableList.of(sourceType)).getSignature().getReturnType();
+                    }
+                    catch (OperatorNotFoundException e) {
+                        throw semanticException(INVALID_PATH, pathNode, e, "invalid operand type (%s) in JSON path arithmetic unary expression: %s", sourceType.getDisplayName(), e.getMessage());
+                    }
+                    types.put(PathNodeRef.of(node), resultType);
+                    return resultType;
+                }
+            }
+
+            return null;
```

Comment:
Would it work to return `UNKNOWN` instead of `null` as a marker that we cannot determine the type?

Reply:
No, `UNKNOWN` is a certain type.
In fact, the semantics of json path is such that `UNKNOWN` does not appear in the `JsonPathAnalyzer`. So, technically we could use it as a marker, but that would be confusing.